lib/ts/recipe/oauth2/types.ts

```diff
@@ -0,0 +1,298 @@
+    >;
+
+    verifyAccessToken(input: {
+        clientId: string;
+        accessToken: string;
+    }): Promise<{
+        sessionHandle?: string;
+        scopes: string[];
+        timeCreated: number;
+        lastUpdated: number;
+        timeAccessTokenExpires: number;
+        timeRefreshTokenExpires?: number;
+    }>;
+
+    getTokenInfo(input: {
+        clientId: string;
+        refreshToken?: string;
+        authCode?: string;
+    }): Promise<{
+        sessionHandle?: string;
+        scopes: string[];
+        timeCreated: number;
+        lastUpdated: number;
+        timeAccessTokenExpires: number;
+        timeRefreshTokenExpires?: number;
+    }>;
+
+    revokeToken(input: { clientId: string } | { sessionHandle: string } | { accessToken: string }): Promise<void>;
+    revokeAuthCodes(input: { clientId: string } | { sessionHandle: string }): Promise<void>;
+
+    // Customization points called by createTokens&createAuthCode
+    buildAccessToken(input: {
+        clientId: string;
+        userId?: string;
+        sessionHandle?: string;
+        scopes: string[];
+        userContext: any;
+    }): Promise<{ payload?: JSONObject; lifetimeInSecs?: number }>; // Returning undefined for these props means we use the default from Core
+    buildIdToken(input: {
+        clientId: string;
+        userId?: string;
+        sessionHandle?: string;
+        scopes: string[];
+        userContext: any;
+    }): Promise<{ payload?: JSONObject; lifetimeInSecs?: number }>; // Returning undefined for these props means we use the defaults from BE SDKs
+    getRefreshTokenLifetime(): Promise<number | undefined>; // Returning undefined means we use the default from Core
+
+    validateScopes(input: {
+        clientId: string;
+        userId?: string;
+        sessionHandle?: string;
+        scopes: string[];
+        userContext: any;
+    }): Promise<
+        | { status: "OK"; grantedScopes: string[] }
+        | { status: "INVALID_SCOPE_ERROR"; message: string } // Returning this signifies a client error and should result in a redirection back to the client with the "invalid_scope" error
+        | { status: "ACCESS_DENIED_ERROR"; message: string } // This means we need to redirect to the auth frontend (i.e.: there is a claim validation issue we can resolve there)
+    >;
+
+    shouldIssueRefreshToken(input: {
+        clientId: string;
+        userId?: string;
+        sessionHandle?: string;
+        scopes: string[];
+        userContext: any;
+    }): Promise<boolean>;
+};
+
+export type ResponseType = "code" | "id_token" | "token"; // We do not support "token", but it's a possible input value.
+export type CodeChallengeMethod = "S256";
+
+export type GrantType = "authorization_code" | "client_credentials" | "refresh_token" | "password"; // We do not support "password", but it's a possible input value.
```

Comment:
If we do not support password, what happens if the user gives that?

Reply:
we return w/ `unsupported_grant_type`